src/librbd/cache/file/SyncFile.cc

```diff
@@ -0,0 +1,269 @@
+void SyncFile<I>::close(Context *on_finish) {
+  assert(m_fd >= 0);
+  while (true) {
+    int r = ::close(m_fd);
+    if (r == -1) {
+      r = -errno;
+      if (r == -EINTR) {
+        continue;
+      }
+      on_finish->complete(r);
+      return;
+    }
+    break;
+  }
+
+  m_fd = -1;
+  on_finish->complete(0);
+}
+
+template <typename I>
+void SyncFile<I>::read(uint64_t offset, uint64_t length, ceph::bufferlist *bl,
+                      Context *on_finish) {
+  on_finish->complete(read(offset, length, bl));
+}
+
+template <typename I>
+void SyncFile<I>::write(uint64_t offset, ceph::bufferlist &&bl,
+                       bool fdatasync, Context *on_finish) {
+  on_finish->complete(write(offset, bl, fdatasync));
+  //on_finish->complete(0);
+}
+
+template <typename I>
+void SyncFile<I>::discard(uint64_t offset, uint64_t length, bool fdatasync,
+                         Context *on_finish) {
+  on_finish->complete(discard(offset, length, fdatasync));
+}
+
+template <typename I>
+void SyncFile<I>::truncate(uint64_t length, bool fdatasync, Context *on_finish) {
+  on_finish->complete(truncate(length, fdatasync));
+}
+
+template <typename I>
+void SyncFile<I>::fsync(Context *on_finish) {
+  int r = ::fsync(m_fd);
+  if (r == -1) {
+    r = -errno;
+    on_finish->complete(r);
+    return;
+  }
+  on_finish->complete(0);
+}
+
+template <typename I>
+void SyncFile<I>::fdatasync(Context *on_finish) {
+  on_finish->complete(fdatasync());
+}
+
+template <typename I>
+int SyncFile<I>::write(uint64_t offset, const ceph::bufferlist &bl,
+                      bool sync) {
+  sync = false;
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "offset=" << offset << ", "
+                 << "length=" << bl.length() << dendl;
+
+  //int r = bl.write_fd(m_fd, offset);
+  //todo: bl is not aligned in one 4096 block
+  char *aligned_buffer = (char*)aligned_alloc(4096, sizeof(char)*4096);
+  int r;
+  /*r = posix_memalign((void**)&aligned_buffer, sizeof(char)*4096, 1);
+  if( r < 0 ){
+    return -r;
+  }*/
+  bl.copy( 0, bl.length(), aligned_buffer );
+  r = pwrite(m_fd, aligned_buffer, 4096, offset); 
+  if (r < 0) {
+    free(aligned_buffer);
+    return -r;
+  }
+
+  if (sync) {
+    r = fdatasync();
+  }
+  free(aligned_buffer);
+  return r;
+}
+
+template <typename I>
+int SyncFile<I>::read(uint64_t offset, uint64_t length, ceph::bufferlist *bl) {
+
+  CephContext *cct = m_image_ctx.cct;
+  bufferptr bp = buffer::create(length);
+  bl->push_back(bp);
+
+  int r = 0;
+  char *buffer = reinterpret_cast<char *>(bp.c_str());
+  char *aligned_buffer = (char*)aligned_alloc(4096, sizeof(char)*4096);
+  /*char *aligned_buffer;
+  r = posix_memalign((void**)&aligned_buffer, sizeof(char)*4096, 1);
+  if( r < 0 ){
+    return -r;
+  }*/
+  uint64_t left = length;
+  ssize_t ret_val;
+  uint64_t count = 0;
+  uint64_t cpy_length;
+  do {
+    //todo: offset is not aligned.
+    ret_val = pread64(m_fd, aligned_buffer, 4096, offset);
```

Comment:
Also probably a good case for reusing / extending the existing ```bufferlist::read_fd``` (i.e. add support for offsets)